### DISCUSSION TITLE
Implement property initializers for immutable types

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#322] [General]
_From @dsaf on February 4, 2015 9:34_

Immutable types are a best practice nowadays. However, when a C# class has a lot of properties it means ctor will have lots of parameters. Property initializers are easier to read and understand as well.
Can we have a way of initializing properties on immutable types? I have read about C# 6 auto-property initializers and unless I am mistaken they only help with compile-time constants.

PS: please don't do this by introducing new types (e.g. why have some sort of 'record' when we already have classes and structs?).

_Copied from original issue: dotnet/roslyn#229_

**Comment by gafter** [comment]
_From @SolalPirelli on February 4, 2015 11:54_

You can already assign values to get-only properties from the constructor in C# 6. ([source](https://roslyn.codeplex.com/discussions/568824))

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 12:1_

@SolalPirelli

This is not what I am asking. I want this:

``` csharp
var student = new Student(new Guid())
{
   Name = "John",
   Address = GetAddressFromDb(),
   Age = minAge + 20
}
```

All properties in the sample above are immutable (readonly).

**Comment by gafter** [comment]
_From @AlgorithmsAreCool on February 4, 2015 13:6_

So you would like to see object initializers gain the ability to write to getter only properties? Like this?

```
class Foo
{
    public int Bar { get; }
}
...

Foo myFoo = new Foo() { 
    Bar = 42
};
```

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 13:17_

@AlgorithmsAreCool

Yes, exactly. Otherwise the number of constructor parameters is starting to affect the choice of whether to make something immutable or not. Which is not great at all.

Furthermore, it would need to be reflection-friendly to make sure deserialisation (e.g. from JSON) is still possible.

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 13:22_

This is already done for anonymous types (although for a different reason):

http://stackoverflow.com/questions/1089406/why-are-the-properties-of-anonymous-types-in-c-sharp-read-only

http://blogs.msdn.com/b/sreekarc/archive/2007/04/03/immutable-the-new-anonymous-type.aspx

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 13:26_

Maybe 'readonly' keyword could be re-used to mark the whole type:

``` csharp
 //All properties and fields have to be declared as readonly or const and be initialised explicitly.
readonly class Student { ... }
```

This post shares this idea as well:

http://madprops.org/blog/immutability-in-csharp-5

I would prefer to see this rather than 'record' which feels kind of alien: name doesn't represent the purpose, syntax is heavily F#-ish etc.

**Comment by gafter** [comment]
_From @AlgorithmsAreCool on February 4, 2015 13:41_

Have you looked at #159 ? It proposes a new immutable type syntax. Your initializer idea might add to that discussion.

**Comment by gafter** [comment]
_From @SolalPirelli on February 4, 2015 15:53_

This can be done with record types (#206) and named arguments.  
For instance, if you have a `class Point(int X, int Y)`, you could instantiate it with `new Point(X = 0, Y = 42)`.

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 16:36_

@SolalPirelli

Yes, but I don't want a "...type whose semantic meaning is described by the shape of the data..." I just want an immutable class. Record types will be created for a different reason (to facilitate pattern matching).

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 16:40_

@AlgorithmsAreCool

Thanks for linking.

**Comment by gafter** [comment]
_From @agocke on February 4, 2015 19:5_

@dsaf 

> Record types will be created for a different reason (to facilitate pattern matching).

So? If it does what you want, why do you care what it was created for? :)

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 20:4_

@agocke 

When other people will be reading my code I want them to say "OK, so Student is not going to change it's state after instantiation, so I don't have to worry about that, great." and not "OK, so Student isn't representing a domain object but is rather describing a semantic meaning of some data that can be shaped in different ways... I better check what is that data, what are we matching it for, and where is the actual Student domain object, so that I could rename it to something else to avoid confusion with that pattern matching record type...".

**Comment by gafter** [comment]
_From @agocke on February 4, 2015 20:9_

@dsaf 

Don't worry -- no one's going to say that. Welcome to the language design. :)

We discuss the backing state machines of iterators and the semantic structure of records in pattern matching so someone using the code can just say "Oh, that's an IEnumerable+yield" or "Oh, that's an immutable class I can use pattern matching on." Only the designers think about how the sausage is made.

**Comment by gafter** [comment]
_From @SolalPirelli on February 4, 2015 20:14_

@dsaf: Scala has `case class`, which is close to #206, and case classes are also used as a convenient way to do what you want here, even when no pattern matching is used.

**Comment by gafter** [comment]
_From @ashmind on February 4, 2015 21:46_

Idea based on #159 for the type syntax.

Let's start with:

``` csharp
public immutable class Person {
    public Person(string name) {
        Name = name;
    }

    public string Name { get; }
    public DateTime Birthday { get; }
    ...
}
```

We want to do `var person = new Person("X") { Birthday = birthday }`.

What if 

``` csharp
public Person(string name) {
    Name = name;
}
```

actually generated

``` csharp
public Person(string name, [GeneratedForImmutable] Generated_PersonPropertyBag properties)  {
    Name = name;
    // generated
    if (properties.NameSet) Name = properties.Name;
    if (properties.BirthdaySet) Birthday = properties.Birthday;
}
```

and then `var person = new Person("X") { Birthday = birthday }` generated

``` csharp
var person = new Person("X", new Generated_PersonPropertyBag {
    BirthdaySet = true,
    Birthday = birthday
})
```

?

**Comment by gafter** [comment]
_From @dsaf on February 4, 2015 22:5_

@ashmind #159 is a bit hardcore, frankly I would appreciate even some simple syntax sugar for read-only properties, not a 100% deep immutable types. What if I want this:

``` csharp
public class Sample
{
   public List<string> SomeProperyOne { get; }
   public List<string> SomeProperyTwo { get; }
   public List<string> SomeProperyThree { get; }
   public List<string> SomeProperyFour { get; }
   public List<string> SomeProperyFive { get; }
}
```

**Comment by gafter** [comment]
_From @ashmind on February 4, 2015 22:41_

@dsaf
But those are lists, you can initialize them in the current C#?
E.g. `new Sample { SomePropertyOne = { element1, element2, etc } }`

(Given that they are not-null, but they should be!)

**Comment by gafter** [comment]
_From @dsaf on February 5, 2015 18:5_

@ashmind 

The point is that those properties are all mutable themselves, but there is no reason why they cannot be read-only to make the host class "shallow-mutable". If they will be readonly I will not be able to initialize them outside of constructor. #159 would forbid using types such as List<> because they are mutable.

**Comment by gafter** [comment]
_From @ashmind on February 5, 2015 19:3_

@dsaf Sorry, but I don't understand. You can initialize `List<T>` properties in current C# outside of the constructor, even if the property setter is not available (it compiles to `get`+`Add`).

The only problem I see if the class is deep-readonly, but that's similar to #159.

**Comment by gafter** [comment]
_From @dsaf on February 5, 2015 20:18_

@ashmind List<> was just an example, it could be any mutable type. There is a difference between a read-only and a completely immutable property.

There is no way to do the following in current or proposed C#: have read-only auto-properties that are not necessarily of an immutable type that would be initialized using property initializes. Please refer to my first sample.

**Comment by gafter** [comment]
_From @thomaslevesque on February 7, 2015 13:23_

This feature could easily be done by introducing the notion of a builder class associated to an immutable class.

```
[BuilderType(typeof(FooBuilder))]
class Foo // immutable
{
    public int Id { get; }
    public string Name { get; }

    public Foo(int id, string name)
    {
        Id = id;
        Name = name;
    }
}

class FooBuilder : IBuilder<Foo>
{
    public int Id { get; set; }
    public string Name { get; set; }
    public Foo Create() => new Foo(Id, Name);
}
```

With this approach, the compiler would transform this:

```
var foo = new Foo { Id = 42, Name = "Joe" };
```

into this:

```
var foo = new FooBuilder { Id = 42, Name = "Joe" }.Create();
```

Of course, it requires the developer to write the builder class, but I think this approach is the most flexible. It would also probably be quite easy to implement.

**Comment by gafter** [comment]
_From @dsaf on February 7, 2015 13:51_

@thomaslevesque the idea is to have less boilerplate code while keeping it readable. I would rather misuse misnamed 'record' classes than have an extra class per every immutable class. Let's move towards Scala not Java. But thank you for being constructive!

**Comment by gafter** [comment]
_From @Joe4evr on March 2, 2015 16:32_

So I came up with a possible solution to this. Should I post it in this thread or make a new proposal?

**Comment by gafter** [comment]
@Joe4evr What is your idea?

**Comment by gafter** [comment]
_From @agocke on March 2, 2015 19:0_

@Joe4evr @gafter He also happens to have a marvelous proof that his proposal is optimal, but unfortunately it's too long to fit into a Github comment. ;)

**Comment by gafter** [comment]
_From @Joe4evr on March 2, 2015 19:28_

@agocke lol, I'm not going to Fermat-levels here, but I feel this is much better suited to the language than that inner struct implementation that was suggested 3 weeks ago.

@gafter It comes down to implicitly generating a constructor, just like the parameter-less default constructor that's already made if one isn't provided. This particular one will have all of the getter-only auto-properties as parameters, and when creating an instance with Object Initialization, the compiler looks at what properties are read-only and puts those values into a constructor call. Example:

``` c#
public class Foo
{
    public string Bar { get; }

    //This constructor would be implicitly generated
    //Properties with regular setters would not have an associated parameter here
    //Parameters could probably be optional, see below
    public Foo(string _Bar)
    {
        this.Bar = _Bar;
    }
}

//Elsewhere....
Foo myFoo = new Foo
{
    Bar = "This is my bar. There are many like it, but this one is mine."
};
//would compile into the following (named parameter used for clarity)
Foo myFoo = new Foo(_Bar: "This is my bar. There are many like it, but this one is mine.");
```

Since the compiler would know that Bar would correspond to the _Bar parameter in the implicitly generated constructor, it'll move the value into the constructor call in the right place. The rest is just building off of already existing code generation.

Of course, this begs the question: Should this implicit constructor be made in addition to the normal default constructor, or should it **be** the default constructor for classes with getter-only auto-properties? I'm thinking it could be the latter, if the parameters are made optional and use the `default` of that parameter's type when not specified (this point is mostly for compatibility reasons).

Thoughts?

**Comment by gafter** [comment]
_From @ashmind on March 2, 2015 20:2_

@Joe4evr seems similar to my struct idea. It is simpler, but in your case any new property will become a binary breaking change -- which I tried to avoid.

**Comment by gafter** [comment]
_From @agocke on March 2, 2015 20:2_

@Joe4evr This looks like the same backing implementation as the pattern matching `record` class, the difference being that in pattern matching you use the constructor directly rather than setting fields. The `record` also has much simpler and more succinct syntax IMHO. That is, this is the in-source definition of your example:

```
record class Foo(string Bar);
```

And you can declare one simply by `new Foo("Bar text blah blah")`. Seems like a lot less work to me.

@gafter Please feel free to correct me if I've misunderstood the most recent spec. I don't remember if we finally got rid of that pesky `record` modifier.

**Comment by gafter** [comment]
@Joe4evr @agocke There is no longer any need for the `record` modifier in the pattern-matching spec.

The intent of the current pattern/records proposal is that the compiler will generate the constructor based on the properties given in the class header. You could put other properties in the body to get additional read-write properties that do not have corresponding constructor parameters.

@Joe4evr If I understand your proposal, the main point is that you could use the object initializer syntax and the compiler will recognize it as constructor parameters instead of separate property assignments. I think that is an attractive idea, and definitely something we're considering.

**Comment by gafter** [comment]
_From @thomaslevesque on March 2, 2015 20:30_

It's an attractive idea, but it comes with a few issues:
- as mentioned by @ashmind, adding properties breaks binary compatibility
- when would a constructor be generated? Only when there isn't an explicit one?
- in which order would the constructor parameters be declared? same order as the properties? In this case, changing the order of the properties would be a breaking change at the source level

**Comment by gafter** [comment]
_From @Joe4evr on March 2, 2015 20:30_

@ashmind That's the one thing, yes, but if the parameters are made optional, then the most one would have to do in case of a library update is rebuild against the new version (provided the library author doesn't make consumers do more, but I think at that point, the author probably would've used an explicit constructor in the first place).

@agocke The main thing is that you can't assign to read-only backing fields outside of a constructor. I don't know how record classes work under the hood, but if you can set their fields directly, they don't sound like read-only.

@gafter That's entirely correct.

@thomaslevesque Yes, not exactly binary compatible, unless the compiler could emit additional metadata.

> when would a constructor be generated? Only when there isn't an explicit one?.

That's the idea. If an explicit constructor is provided, then the onus is on the author to make sure all accessible getter-only auto-props can be assigned to.

> in which order would the constructor parameters be declared? same order as the properties? In this case, changing the order of the properties would be a breaking change at the source level

Good question, perhaps alphabetical order should then be generated under the hood. The compiler would then place the parameters in their proper location, just like it already does for named parameters.

(5 edits: Markdown, plz)

**Comment by gafter** [comment]
_From @agocke on March 2, 2015 20:50_

@Joe4evr As I said, they generate a constructor, so the constructor parameters become the field values. The only difference between the mapping in your implementation and records is that you've overridden the object initializer syntax to map to a constructor. Records don't contain an extra layer of indirection -- they just use a constructor directly.

**Comment by gafter** [comment]
_From @ashmind on March 2, 2015 22:12_

@gafter

> There is no longer any need for the record modifier in the pattern-matching spec.

That's interesting -- which issue has the latest discussion?

@Joe4evr
It is not easy to rebuild a library -- e.g. if you have libA and libB on NuGet using different version of libC -- which is expected to be compatible, but isn't. Of course it happens with other things as well, but the less of those are there, the better.

---

Overall I don't see a better way than struct if we want to have an object initializer for an otherwise readonly set of properties. However that's only within concepts of current C# -- e.g. record types and non-readonly access by the compiler could be also a solution.

**Comment by gafter** [comment]
@ashmind The proposal for records and pattern matching is #206 

Also relevant are 'with' expressions as an extension of that... see https://github.com/dotnet/roslyn/issues/206#issuecomment-72793926

**Comment by gafter** [comment]
_From @alrz on October 11, 2015 11:12_

This is also useful in attribute usages like

``` cs
public class FooAttribute : Attribute { public string Property { get; } }

[Foo(Property = "value")]
```

Although, it makes more sense to declare this kind of classes in form of a record type, e.g.

``` cs
class Student(Guid guid : Guid)
{
    public string Name { get; }
}
```

There is one problem with this, what if you don't want it to be initializable outside of the enclosing type? Maybe `public readonly string Name { get; }`, then you can initialize it right away `{ get; } = ...` or in ctor, as you could before.

**Comment by john-cd** [comment]
```csharp
var person = new Person("X") { Birthday = birthday }.
```

The problem with this notation is that it suggests asymmetry between the properties (where in fact the only difference is that X is passed explicitly and Birthday implicitly). Perhaps consider discouraging this via a code analysis suggestion. 

Same issue with records type like

```csharp
class Student(Guid guid : Guid)
{
    public string Name { get; }
}
```

To keep Reflection clean, consider avoiding generated constructors like

```csharp 
public Person(string name, [GeneratedForImmutable] Generated_PersonPropertyBag properties)  {
    Name = name;
    // generated
    if (properties.NameSet) Name = properties.Name;
    if (properties.BirthdaySet) Birthday = properties.Birthday;
}
```
in favor of the clean, symmetric constructor ``public Person(string name, int Birthday)``

and perhaps allow

```csharp
public class C 
{

public C()
 : this(Name: "John", Age: 42)   // refer to autogenerated constructor public C(name, age)
{
}

public string Name { get; }   // readonly properties the constructor is generated from

public int Age { get; }

}

var c = new C {  Name = "Mike", Age = 40  };   // object init syntax
var c1 = new C("Joe", 41);                                 // constructor syntax
var c2 = new C();                 
```

**Comment by dsaf** [comment]
> However, when a C# class has a lot of properties it means ctor will have lots of parameters. Property initializers are easier to read and understand as well.

For the record I don't strongly feel about this anymore. It's not a problem to use constructors and named arguments. Beyond that there is no guarantee that a type is immutable anyway.

Besides there would still be a "problem" with immutable collection syntax:

http://stackoverflow.com/questions/16882136/can-i-initialize-a-bcl-immutable-collection-using-braces

>> Is there a similar syntax that can be used with BCL immutable collections?

>Not as far as I'm aware - unfortunately both object and collection initializers rely on mutability. In that respect it's a shame that the language isn't designed such that if there's an Add method with a return value, that can be used as an intermediate value. (That wouldn't help with setting properties in object initializers, mind you, and I prefer the name Plus to Add to make the semantics clearer.)

**Comment by MgSam** [comment]
@gafter I hope you used a script to do all that- either way your dedication is admirable!

**Comment by JendaPerl** [comment]
I don't think it would be a good idea to combine an actual constructor and initializers. The constructor left the struct in a consistent state and the readonly is a promise you would not change that. So IMHO
`new Foo(some, params) {andsome = initializers}` should not be allowed.
On the other hand if there is no constructor and no need for one then having to type out a constructor with all the properties that doesn't do anything else than set the properties to the passed values and keep it updated is tedious and should not be necessary.

I think the safe and still helpful solution would be to allow initializers only if there is no constructor in the readonly struct/class definition.
```C#
 public readonly struct Foo { 
   public readonly int X {get;}
   public readonly int Y {get;}
 }
 ...
 new Foo{ X = 5, Y = 15}, new Foo{ X = 99}
```
would thus become
```C#
 public reaonly struct Foo {
   public Foo(int x = default(int), int y = default(int)) { this.X = x; this.Y = y; }
   public readonly int X {get;}
   public readonly int Y {get;}
 }
 ...
 new Foo(x: 5, y: 15), new Foo(x: 99)
```

**Comment by Joe4evr** [comment]
> So IMHO `new Foo(some, params) {andsome = initializers}` should not be allowed.

Well, I guess it sucks to be you, because [that scenario has always been legal](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABAjANgAJsAmQgYQFgAoAbxoEg8jsAWQgWQAoBKGwgYXrVBowmACGUQnAD2swgF5CAOwgB3QgDF5XViSSF9fEWIHCzZgMpLCAIi1RZAW0IwVMYDAkAbGAC9oO35LAF8AbhCBUJoY6hpsAGZiMh1ZGgtBJOJcAAZCG1pCAHMIYHDCAGcyirjRbPdgQgANIRKawjqs5MbCAE020vLOmijiZLSuXoAPQ16EE0slAD5CLmbDPp5bLlnCBcjqUKA).

**Comment by JendaPerl** [comment]
@Joe4evr In case of readonly structs? I don't think so.

**Comment by HaloFour** [comment]
@JendaPerl 

That syntax is legal anywhere else initializers are permitted, though.  It would be inconsistent to not allow that here.

Either way, initialization via generated constructors is inherently fragile.  Something which should be innocuous like reordering the members will result in breaking binary changes due to the regenerated constructor.

Last I recall the team was looking at instead generating a mutable struct builder and supporting that for readonly initialization.

**Comment by jnm2** [comment]
@JendaPerl Technically, yes. `readonly` only means the struct's storage area is immutable, not that its methods (including property accessors) are pure or that it doesn't mutate things which are referenced from its storage area.

https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABAjANgAJsAmQgYQFgAoAbxoEg8jsAWQgWQAoBKGwgYXrVBowmACGUQnAD2swgF5CAOwgB3QgDF5XViT4ixA4ceMAhKQDkJAWwhLCAIi1RZtwjBUxgMCQBsYAC9oJ34zAF8AbnCBCJp46hpsAGZCKAgJFFkVfwRCAGdgKABXAGNgbXkaU0EAB1hJYAcMrJy8wktpACMpGKMBWOI0vAAGTus7B1pCAHMISsUAPkJeqAA6G3sowoWlFbXNqcdJfxKIHcSh1KrZLi9KgA9DUVrjNcc1TS7eftFEq7UG6kCZQGrXEa4cZbaZzBY7ArwwgAoA===

**Comment by JendaPerl** [comment]
> @JendaPerl
> That syntax is legal anywhere else initializers are permitted, though. It would be inconsistent to not allow that here.

It's not allowed now for readonly structs and IMHO it should not be. Either constructor or initializers, not both. If I need to set things up so that the whole structure is in a consistent state, I use a constructor and as I set the structure to readonly I do not expect the state to change. If the structure is just a dumb data store, then having to write a constructor is tedious and error prone.

> Either way, initialization via generated constructors is inherently fragile. Something which should be innocuous like reordering the members will result in breaking binary changes due to the regenerated constructor.

As the constructor would be called "behind the scenes" and using named parameters, there is no reason the parameters could not be sorted alphabetically instead of being left in the "order of appearance". Then you'd only have a breaking binary change if you added or removed properties or changed their types.

> Last I recall the team was looking at instead generating a mutable struct builder and supporting that for readonly initialization.

Whatever works :-)
I'd just really like it to work in Linq over Entity Framework.

**Comment by jnm2** [comment]
@JendaPerl 

>It's not allowed now for readonly structs

It has always been allowed for readonly structs. Did you see the [sharplab example](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABAjANgAJsAmQgYQFgAoAbxoEg8jsAWQgWQAoBKGwgYXrVBowmACGUQnAD2swgF5CAOwgB3QgCUIElAHkVAGwQBlYFACuAY2BdWJPiLEDhLlwCEpAOQkBbCCVCACIAMShZP0IYFRhgGAkjGAAvaGD+dwBfAG4MgUyaAuoabABmQihdFFljBEIAZwsbYG0qwxNzK1saN0EAB1hJYEDKvRqTQi9pACMpXOcBPOJyvAAGSZ9/QNpCAHMIFsUAPkJZqAA6XwDshoOlE7PLraDJI0sIG6Klsta9drMmrYuDEWgAPJyiXouM5BNSaKa8eaiIpfag/UgbKA9b4rXDrK7bPYHG71YmEFFAA) I linked? This was legal code the day the `readonly struct` feature shipped:

```cs
static class C
{
    static void M()
    {
        var foo = new ReadOnlyStruct(42)
        {
            BarName = "From initializer"
        };
    }
}

public readonly struct ReadOnlyStruct
{
    private readonly Bar bar;
    
    public string BarName { get => bar.Name; set => bar.Name = value; }

    public ReadOnlyStruct(int x)
    {
        bar = new Bar();
    }
}

public class Bar
{
    public string Name { get; set; }
}
```

The reason that this is consistent is that `readonly` has a very specific meaning. `readonly` only means the struct's *storage area* is immutable, not that its methods (including property accessors) are pure or that it doesn't mutate things which are referenced from its storage area.

**Comment by JendaPerl** [comment]
> ```cs
> static class C
> {
>     static void M()
>     {
>         var foo = new ReadOnlyStruct(42)
>         {
>             BarName = "From initializer"
>         };
>     }
> }
> 
> public readonly struct ReadOnlyStruct
> {
>     private readonly Bar bar;
>     
>     public string BarName { get => bar.Name; set => bar.Name = value; }
> 
>     public ReadOnlyStruct(int x)
>     {
>         bar = new Bar();
>     }
> }
> 
> public class Bar
> {
>     public string Name { get; set; }
> }
> ```

And this is useful how?!? You bend backwards to add a non-readonly property to a struct marked readonly and then and only then you are able to mention it in an initializer? COOL!

The point is to be able to INITIALIZE the readonly properties as I instantiate the struct, not whether you can circumvent the readonly mark on a struct to add something that's not readonly.

**Comment by KubaO** [comment]
As much as the proposed feature would be handy, I think that being explicit via constructors is more helpful. It seems to help with design of the class in question: often the initializer syntax is an "organic growth" on top of well thought-out constructors, and indicates a need to refactor things. For classes that have a clear and sensible™ need for combined constructor + initializer syntax, the property bag / builder type pattern should work well. Those should be few and far between. So, while I initially thought that this was a very good idea, I don't think so anymore.

**Comment by JendaPerl** [comment]
Being explicit is insanely chatty in case of the majority of classes whose sole purpose of existence is to hold some data. There's nothing to design there, I just want to pass around the data I just read from somewhere.
I do see very little need for combined constructor + initializer in any case and it would defeat the purpose in case of readonly structures, but begin forced to create the one and only ever constructor that mentions every single property is simply silly.
